riot/riot.ml

```diff
@@ -40,6 +40,11 @@ let run ?(rnd = Random.State.make_self_init ()) ?workers main =
   Log.debug (fun f -> f "Riot runtime shutdown");
   Stdlib.exit pool.status
 
+let run_with_status ?(rnd = Random.State.make_self_init ()) ?workers main =
+  run ~rnd ?workers (fun _ ->
+      let status = main () |> Result.get_ok in
+      shutdown ~status ())
```

Comment:
**suggestion**. This wrapping looks good but we can improve it a little bit since if `main ()` returns an error we will have a runtime exception and the application will crash without shutting down the engine correctly.

The right way of handling this would be to pattern-match on `main ()` and handle both the `Ok status` case and the `Error reason` case.

For the Ok case, your current code looks good.

For the error case, we will need to ask the user how to handle it. We can this by also accepting a new parameter like `on_error` which by default is defined as a function that handles `` `Msg string `` errors:

```ocaml
let default_on_error (`Msg reason) = 
  (* ... print error msg ... *)
```

This would allow a user to also pass in their own function if they want to have other kinds of errors handled.

What's important is that _after_ calling their function, we always call `shutdown ()`.

Reply:
I implemented this but I don't know what to do about cases where the error is not `` `Msg of string `` (currently it just says the error is unsupported for printing).

Reply:
Left you a few comments that should help with this :)